0003-siren-prototype-accept-decline.adoc

```diff
@@ -0,0 +1,88 @@
+= Siren Prototype for comit-rs HTTP API - Accept/Decline actions
+
+* Authors: Franck Royer
+* Date: 2019-05-02
+
+Tracking issue: https://github.com/comit-network/comit-rs/issues/917[comit-rs#917]
+
+== Context
+
+In order to move forward easily with Bobtimus prime we should define the siren json document of how a IN_PROGRESS response look like in each state of accept and decline, i.e. when you call GET /swaps/ for Bob.
+
+Siren repository: https://github.com/kevinswiber/siren/
+
+The prototype must conform to https://github.com/kevinswiber/siren/blob/master/siren.schema.json[Siren JSON schema]
+
+== Research
+
+.The prototype
+[source,json]
+----
+{
+  "class": ["swaps"],
+  "entities": [
+    {
+      "class": ["swap"],
+      "rel": [],
+      "properties": {
+        "role": "Alice",
+        "protocol": "rfc003",
+        "status": "IN_PROGRESS",
+        "parameters": {
+          "alpha_asset": {
+            "name": "bitcoin",
+            "quantity": "100000000"
+          },
+          "alpha_ledger": {
+            "name": "bitcoin",
+            "network": "regtest"
+          },
+          "beta_asset": {
+            "name": "ether",
+            "quantity": "10000000000000000000"
+          },
+          "beta_ledger": {
+            "name": "ethereum",
+            "network": "regtest"
+          }
+        }
+      },
+      "links": [
+        {"rel": ["self"], "href": "/swaps/rfc003/399e8ff5-9729-479e-aad8-49b03f8fc5d5"},
+        {
+          "rel": ["human-protocol-spec"],
+          "href": "https://github.com/comit-network/RFCs/blob/master/RFC-003-SWAP-Basic.md"
+        }
+      ],
+      "actions": [
+        {
+          "name": "accept",
+          "title": "Accept",
+          "method": "POST",
+          "href": "/swaps/rfc003/399e8ff5-9729-479e-aad8-49b03f8fc5d5/accept",
+          "type": "application/json",
+          "class": ["comit-action"],
+          "fields": [
+            {
+              "name": "beta_ledger_refund_identity", "title": "Refund address on Ethereum", "type": "text",
+              "class": ["address", "ethereum"]
```

Comment:
:x: (to enforce discussion)

@LLFourn voiced some concerns at some point that this should maybe be a single value `ethereum-address` instead of two. What do you guys think?

Reply:
yeah at first I was surprising to see `["address", "ethereum"]` in the draft as I was expecting `address-ethereum`.

I have no strong opinion on this. I think it makes sense to separate because at the end, when writing `address-ethereum` we are creating our own format to express the fact that it is an `address` on the `ethereum` network.
Having it separated makes slightly more sense to me.

Reply:
The thing that makes separation a bit tricky is, how to you use these values in automated system?

You basically have to try and match against a known set of values to figure out, which classes are present because all you have is a list of strings. The logic would be like:

```
if classes.contains("ethereum") && classes.contains("address") {
    return metamask.getAddress()
}
```

Thoughts @LLFourn, @bonomat ?

Reply:
it'd probably be more nested than that:

```
if classes.contains("ethereum") {
  if classes.contains("address") {
    return metamask.getAddress()
  } else if classes.contain("broadcast-transaction") {
  // ...
  }
}
```

no?

Reply:
I am not sure why any field would ever be tagged with a class `broadcast-transaction`?

Whether or not you nest it is a matter of taste I guess. I'd prefer not nesting it to keep things local and modular, i.e. detecting that a field is an ethereum address has little to do with detecting the field which represents the fee per byte on bitcoin. Nested `if` statements convey that requirements are connected, whereas if they are not, I'd prefer the code to not suggest that.

Reply:
Unfortunately siren does not give a lot of documentation about the thoughts behind this field besides: `Describes the nature of an entity's content based on the current representation. Possible values are implementation-dependent and should be documented. MUST be an array of strings. Optional` 

I googled quite a bit and there is not much usage of siren out there neither. 
However, I have the feeling that the *siren-way* of doing this would be indeed `["address", "ethereum"]`. 

The way how I would read it is that this field is of type `address` of type `ethereum`, or expressed in code: 
```
let ethereum_types = implementation_detail_types.get("ethereum")
let address_type = ethereum_types.get("address")
```

If we look at the [examples](https://github.com/kevinswiber/siren):
``` [...] "class": [ "info", "customer" ], [..] ```

this could be interpreted as: 
```
let customer_types = implementation_detail_types.get("customer")
let info_type = customer_types.get("info")
```

There can be some well known types which allow some kind of short forms and tell you directly what type it may be, e.g. 
```[...] "class": [ "items", "collection" ], [...]```

----

... on the other hand, to simplify things, I would go for the most specific version, i.e. `ethereum-address`  😬

Reply:
Let's list the possible values to understand what would be the best solution. I think it does not matter in any case.

Reply:
@bonomat I don't really understand where you are going with your comment :thinking: 

From my understanding, `class` is a way of _tagging_ entities, fields and actions with domain-specific data. This serves the purpose of identifying and finding specific actions, fields or entities within the response.
The client could would therefore:

- first take all actions in the response and look, which of them are tagged with `comit-action`
- those it would render as buttons in the list
- upon clicking an action, the client can inspect the `fields` of this action and, if present, render a form for all fields
- to enhance the user experience, it can search for known values in the `class` array to invoke special behaviour for some fields, like replacing the input field with an `EthereumAddressTextField` (which has support for metamask)

The question is therefore how "modular" we want this detection to be:

Do we want to repeat ourselves in defining single-token classes like:
- ethereum-address
- bitcoin-address
- bitcoin-fee
- ...

or do we want to split those tags into:

- bitcoin
- ethereum
- address
- fee
- ...

Splitting them up would make them more composable, which might be an advantage in the future.
On the other hand, the first option has the advantage of making invalid combinations of classes impossible.
However, invalid combinations are actually not so much of a issue since the client is only going to invoke special behaviour upon known combinations, hence and invalid combination will not break clients.

Reply:
I think this thread is already to long for the value of the decision that can always be easily reverted.

Reply:
@thomaseizinger :
Along the lines this summarizes what I meant: 

> * to enhance the user experience, it can search for **known** values in the `class` array to invoke special behaviour for some fields, like replacing the input field with an `EthereumAddressTextField` (which has support for metamask)
...
> Splitting them up would make them more composable, which might be an advantage in the future.
> On the other hand, the first option has the advantage of making invalid combinations of classes impossible.

@D4nte 
> After coding it I prefer 2 classes to more easily transfer the responsibility of getting the data:

Well then let's go for this approach.

Reply:
Works for me!